[rllib] Behavior Cloning

royf opened the pull request:
Implements the Behavior Cloning (BC) algorithm in RLlib.

To run it:
1. Train a supervisor with any algorithm.
2. Rollout the trained supervisor to demonstrate the behavior to clone:
`rollout.py <checkpoint_path> --run <training_algorithm> --env <env_name> --steps <demonstration_steps> --out <rollouts_path>`
3. Train from these demonstrations:
`train.py --run BC --env <env_name> --config '{"dataset_path": <rollouts_path>}'`

Tested on FrozenLake-v0 with 100K demonstration steps of a DQN supervisor.

AmplabJenkins commented:
Can one of the admins verify this patch?

richardliaw commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3156/
Test PASSed.

richardliaw commented on python/ray/rllib/a3c/a3c.py:
```diff
@@ -66,6 +66,7 @@
 class A3CAgent(Agent):
     _agent_name = "A3C"
     _default_config = DEFAULT_CONFIG
+    _allow_unknown_configs = True
```
probably want to keep this as `False`

royf replied:
without this I got `Unknown config parameter vf_loss_coeff`

richardliaw replied:
really? `vf_loss_coeff` is defined above in line 40; can you help me reproduce this error?

royf replied:
Nevermind, I guess I mixed incompatible installation and `train.py` revisions.

richardliaw commented on python/ray/rllib/rollout.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
```
I believe this is very similar to `eval.py` - perhaps just modify that instead of introducing a new file?

royf replied:
that's reasonable; `eval.py` is missing `ModelCatalog.get_preprocessor_as_wrapper` though, right?

richardliaw replied:
yeah, the wrapper would be a nice addition

richardliaw commented on python/ray/rllib/optimizers/optimizer.py:
```diff
@@ -7,8 +7,8 @@ class Optimizer(object):
     For example, AsyncOptimizer is used for A3C, and LocalMultiGPUOptimizer is
-    used for PPO. These optimizers are all pluggable however, it is possible
-    to mix as match as needed.
+    used for PPO. These optimizers are all pluggable and it is possible
```
"all pluggable, and"

richardliaw commented on python/ray/rllib/bc/tfpolicy.py:
```diff
@@ -0,0 +1,77 @@
+        # TODO(rliaw): Can consider exposing these parameters
+        self.sess = tf.Session(graph=self.g, config=tf.ConfigProto(
+            intra_op_parallelism_threads=1, inter_op_parallelism_threads=2,
+            gpu_options=tf.GPUOptions(allow_growth=True)))
```
is this the only change in this file compared to `a3c/tfpolicy`? If so, we should just expose these parameters via the default config in A3C and avoid repeating too much code.

royf replied:
the bigger change is removing the `assert` in `__init__`; and I didn't want to create a dependency of BC on A3C; we should refactor the policy to be external to either of these

richardliaw replied:
I see; there is still a dependency on A3C via `a3c/policy.py`. I guess it would be reasonable to duplicate all of the policy code for now, @ericl?

ericl replied:
Duplicating sounds good, there's not a lot.

Btw, shouldn't we combine Policy and TFPolicy into one class? There doesn't seem to be much value in separating them.

richardliaw replied:
Yeah, probably should combine `bc/policy.py` and `bc/tfpolicy.py`. We can leave the `a3c/policy.py` dependency as it's just an interface.

richardliaw commented on python/ray/rllib/bc/policy.py:
```diff
@@ -0,0 +1,50 @@
+
+
+class Policy(TFPolicy):
+    def __init__(self, registry, ob_space, ac_space, config, **kwargs):
```
You don't need to include this (I'm pretty sure)

richardliaw commented on python/ray/rllib/bc/policy.py:
```diff
@@ -0,0 +1,50 @@
+from __future__ import absolute_import
```
In the spirit of reducing repetition, this file could probably be extended from `a3c/shared_model.py` with a modification of the `setup_loss`

royf replied:
indeed, they should both be external to any specific algo; this is one of the issues I wanted to open

richardliaw replied:
yeah, the model/framework-dependent code hasn't been refactored yet.

richardliaw commented on python/ray/rllib/a3c/tfpolicy.py:
```diff
@@ -28,7 +28,7 @@ def __init__(self, registry, ob_space, action_space, config,
             self.initialize()
 
-    def _setup_graph(self):
+    def _setup_graph(self, ob_space, ac_space):
```
This is good, thanks.

richardliaw commented on python/ray/rllib/bc/bc.py:
```diff
@@ -0,0 +1,80 @@
+                total_samples += m["num_samples"]
+                total_loss += m["loss"]
+        result = TrainingResult(
+            episode_reward_mean=total_loss / total_samples,
```
I believe there's a specific field for `loss` separate from `reward` - check `tune/result.py`

richardliaw commented on python/ray/rllib/bc/bc.py:
```diff
@@ -0,0 +1,80 @@
+        metric_lists = [re.get_metrics.remote() for re in self.remote_evaluators]
+        total_samples = 0
+        total_loss = 0
+        for metrics in metric_lists:
```
probably just need to do `for m in ray.get(metric_lists)`

royf replied:
`metric_lists` is a list of lists (one per remove evaluator) of dicts (one per finished batch); this parallels `A3CAgent._fetch_metrics_from_remote_evaluators`

richardliaw replied:
I see

richardliaw commented on python/ray/rllib/bc/experience_dataset.py:
```diff
@@ -0,0 +1,21 @@
+from __future__ import absolute_import
```
We will need to extend the functionality of this class to work on multiple machines.

royf replied:
yea, it's currently loading the dataset from a specified path; we may want to ray.put the dataset, would that be a problem with really big datasets (say up to several 10GB)? what's the alternative?

richardliaw replied:
depends on how much memory you allocate for your instance, but you could always partition the dataset so that the object store acts more as a buffer. I think `ray.put` should be fine.

ericl replied:
Put could be hazardous if the dataset gets evicted at some point. Maybe for now just document that the files should be available on each machine.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3168/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3166/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3167/
Test PASSed.

ericl commented:
Looks good. Should we add a paragraph in `doc/source/rllib.rst` on how to use BC?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3186/
Test PASSed.

royf commented:
Commit addresses remaining comments, except [testing](https://github.com/ray-project/ray/pull/1400#pullrequestreview-87093219) and [documentation](https://github.com/ray-project/ray/pull/1400#issuecomment-355854623).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3257/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3310/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3347/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/3356/
Test PASSed.
